stan/math/rev/mat/functor/algebra_solver.hpp

```diff
@@ -0,0 +1,246 @@
+        Eigen::MatrixXd Jf_y = fy.get_jacobian(value_of(y));
+
+        Jx_y_ = - stan::math::mdivide_left(Jf_x, Jf_y);
+      }
+
+      void chain() {
+        for (int i = 0; i < x_size_; i++)
+          for (int j = 0; j < y_size_; j++)
+            y_[j]->adj_ += theta_[i]->adj_ * Jx_y_(i, j);
+      }
+    };
+
+    /**
+     * Return the solution to the specified system of algebraic
+     * equations given an initial guess, and parameters and data,
+     * which get passed into the algebraic system. The user can
+     * also specify the relative tolerance (xtol in Eigen's code),
+     * the function tolerance, and the maximum number of steps
+     * (maxfev in Eigen's code).
+     *
+     * Throw an exception if the norm of f(x), where f is the
+     * output of the algebraic system and x the proposed solution,
+     * is greater than the function tolerance. We here use the
+     * norm as a metric to measure how far we are from 0.
+     *
+     * @tparam F type of equation system function.
+     * @param[in] f Functor that evaluates the system of equations.
+     * @param[in] x Vector of starting values.
+     * @param[in] y parameter vector for the equation system. The function
+     *            is overloaded to treat y as a vector of doubles or of a
+     *            a template type T.
+     * @param[in] dat continuous data vector for the equation system.
+     * @param[in] dat_int integer data vector for the equation system.
+     * @param[in, out] msgs the print stream for warning messages.
+     * @param[in] relative_tolerance determines the convergence criteria
+     *            for the solution.
+     * @param[in] function_tolerance determines whether roots are acceptable.
+     * @param[in] max_num_steps  maximum number of function evaluations.
+     * @return theta Vector of solutions to the system of equations.
+     */
+    template <typename F>
+    Eigen::VectorXd
+    algebra_solver(const F& f,
+                   const Eigen::VectorXd& x,
+                   const Eigen::VectorXd& y,
+                   const std::vector<double>& dat,
+                   const std::vector<int>& dat_int,
+                   std::ostream* msgs = 0,
+                   double relative_tolerance = 1e-10,
+                   double function_tolerance = 1e-6,
+                   long int max_num_steps = 1e+3) {  // NOLINT(runtime/int)
+      // Check that arguments are valid
+      check_nonzero_size("algebra_solver", "initial guess", x);
+      check_nonzero_size("algebra_solver", "parameter vector", y);
+      for (int i = 0; i < x.size(); i++)  // FIX ME - do these w/o for loop?
+        check_finite("algebra_solver", "initial guess", x(i));
+      for (int i = 0; i < y.size(); i++)
+        check_finite("algebra_solver", "parameter vector", y(i));
+      for (size_t i = 0; i < dat.size(); i++)
+        check_finite("algebra_solver", "continuous data", dat[i]);
+
+      if (relative_tolerance <= 0)
```

Comment:
Good point:
```
      if (relative_tolerance < 0)
        invalid_argument("algebra_solver",
                         "relative_tolerance,", relative_tolerance,
                         "", ", must be greater than or equal to 0");
      if (function_tolerance < 0)
        invalid_argument("algebra_solver",
                         "function_tolerance,", function_tolerance,
                         "", ", must be greater than or equal to 0");
```
Mind you, the tolerance for the ODE integrators are required to be strictly positive -- we may want to discuss revising that too.